torch/_functorch/functional_assertions_helper.py

```diff
@@ -0,0 +1,103 @@
+from collections import defaultdict
+from typing import Callable, Dict
+
+import torch
+import torch._decomp as decomp
+from torch._ops import OpOverload
+import torch._functorch.config as config
+
+aten = torch.ops.aten
+_decompositions: Dict[str, Dict[OpOverload, Callable]] = defaultdict(dict)
+
+
+class DepTokenStateTracker:
+    """
+    A singleton / global context manager which is used to:
+    - Record the global flag that assertion ops functionalization is enabled
+      during compilation if the context is entered.
+    - Record most recent token which is propagated through functional assertion
+      ops.
+    """
+
+    _dep_token = None
+    _entered: bool = False
+
+    def __enter__(self) -> "DepTokenStateTracker":
+        assert (
+            not DepTokenStateTracker._entered
+        ), "Cannot reenter singleton context `DepTokenStateTracker`."
+        self.reset()
+        DepTokenStateTracker._entered = True
+        return self
+
+    def __exit__(self, exc_type, exc_cal, exc_tb) -> None:
+        assert DepTokenStateTracker._entered, (
+            "Cannot exit singleton context `DepTokenStateTracker` which is not "
+            "entered previously."
+        )
+        self.reset()
+        DepTokenStateTracker._entered = False
+        return None
+
+    @classmethod
+    def reset(cls) -> None:
+        cls._dep_token = None
+
+    @classmethod
+    def set_dep_token(cls, dep_token) -> None:
+        assert cls._entered
+        cls._dep_token = dep_token
+
+    @classmethod
+    def get_dep_token(cls):
+        if cls._dep_token is None:
+            cls._dep_token = aten._make_dep_token()
```

Comment:
hmm, I see `_make_dep_token` is in `native_functions.yaml`.

We usually try to avoid adding extra ops to `native_functions.yaml`, unless they need the full power of our codegen (autograd support, eager mode support, possible has different implement for different backends, etc).

Does `_make_dep_token` need any of that? Or is it more specific to `torch.export`. If it is, it might make more sense to register it as a custom operator (this is actually pretty easy to do, since you can register ops from python!). E.g. https://github.com/pytorch/pytorch/blob/ef285faebacec86dff5fbf8a8dca8f2672a8c257/test/test_python_dispatch.py#L331

Reply:
> We usually try to avoid adding extra ops to native_functions.yaml, unless they need the full power of our codegen

I see. This is to great to learn though. And I mostly follow @tugsbayasgalan's guidance for creating aten op.

> it might make more sense to register it as a custom operator

Can definitely play with this. One noob question is that it can create an aten level op right?

Reply:
> One noob question is that it can create an aten level op right?

Hmm, I don't think it'll let you create an op in the "aten" namespace (you could test it out - pretty sure it'll raise an error that aten has already been defined elsewhere).

But if this op is only specific to export anyway, I think it's reasonable to put it in another namespace like "torch.ops.export_ops._make_dep_token", etc

Reply:
I feel like it doesn't necessary need to be export specific since we could also enable assertion functionalization during non-export compile? But anyway, this could be a separate discussion since renaming the op will happen in separate PR anyway.